### PR TITLE
add early_stopping; 

### PR DESCRIPTION
**Pull request by vindolineKis**
add `early_stopping()`(patience,min_delta) into `trainerModel` ; add `optimizer.zero_grad()` to prevent gradient accumulation.

### CODE REVIEW

**Comment by cfenglv** [comment]
Conflict due to not matching with the latest version.

**Comment by cfenglv** [comment]
Delete empty lines and unused package

File: QCircLearning/trainerModel.py [hunk 1]
```diff
@@ -122,24 +150,30 @@ def NN_opt(func, x0, callback=None, **kwargs):
             TrainerModel.simple_model((para_size,)),
         ],
     )
-
+    patience = kwargs.get("patience", 5)
+    min_delta = kwargs.get("min_delta", 0.0)
+    print("Patience: ", patience)
```
Comment (cfenglv) [inline QCircLearning/trainerModel.py:155]: Nothing should be printed unless verbose is True. Same for all the rest

File: QCircLearning/trainerModel.py [hunk 2]
```diff
@@ -99,6 +105,28 @@ def to_minimize_with_grad(x):
 
     return result.x
 
+class EarlyStopping:
```
Comment (cfenglv) [inline QCircLearning/trainerModel.py:108]: Move out to utils.py

File: QCircLearning/trainerModel.py [hunk 3]
```diff
@@ -24,6 +23,13 @@ def forward(self, x, y=None):
         else:
             return pred
 
+    def __str__(self):
+        model_structure = self.model.__str__()
+        return f"TrainerModel(name={self.name}):\n{model_structure}"
```
Comment (cfenglv) [inline QCircLearning/trainerModel.py:28]: Suggestion: directly: "return f"TrainerModel(name={self.name}):\n{self.model}""

File: QCircLearning/trainerModel.py [hunk 4]
```diff
@@ -122,24 +150,30 @@ def NN_opt(func, x0, callback=None, **kwargs):
...
     for model in nn_models:
         print(model)
         sys.stdout.flush()
 
-        optimizer = optim.Adam(model.parameters(), lr=1e-4)
+        optimizer = optim.Adam(model.parameters(), lr=1e-3)
 
         for iteration in range(max_iter):
             res.nit += 1
             print(f"Iteration {iteration + 1}/{max_iter}")
```
Comment (cfenglv) [inline QCircLearning/trainerModel.py:175]: revert to PR #4 version, which verbose control and run id is added

Comment (cfenglv) [reply QCircLearning/trainerModel.py:175]: Same for all the rest

**Comment by vindolineKis** [comment]
related to #9
